### ISSUE TITLE
How to use 64MB secure memory at ls1012a

### ISSUE THREAD

**Issue by mofanv** [OP-TEE/optee_os#2833]
Please help!

I want to use  as much as secure memory on a ls1012a board. I tried the followed methods.

1\. Increased `PGT_CACHE_SIZE` in /optee_os/core/arch/arm/include/mm/pgt_cache.h
   #define PGT_CACHE_SIZE ROUNDUP(CFG_NUM_THREADS **2**, PGT_NUM_PGT_PER_PAGE)
   **changed to:**
   #define PGT_CACHE_SIZE ROUNDUP(CFG_NUM_THREADS  **130**, PGT_NUM_PGT_PER_PAGE)

2\. Decreased `TTBCR_N_VALUE` in /optee_os/core/arch/arm/mm/core_mmu_v7.c to the smallest possible
    #define TTBCR_N_VALUE		**7**
    **changed to:**
    #define TTBCR_N_VALUE		**5**

3\. Increased `CFG_SHMEM_SIZE` in /optee_os/core/arch/arm/plat-ls/conf.mk
    CFG_SHMEM_SIZE ?= **0X00200000**
    **changed to:**
    CFG_SHMEM_SIZE ?= **0X00400000**

4\. Tried to change memory size in user_ta_header_defines.h to:
    #define TA_STACK_SIZE               (1 * 1024 * 1024)
    #define TA_DATA_SIZE                (10 * 1024 * 1024)

I find the definition of `CFG_TEE_OS_DRAM0_SIZE` is **0x4000000** for `plat-ls`, which is **64MB**. So the `CFG_TZDRAM_SIZE` is `CFG_TEE_OS_DRAM0_SIZE - CFG_SHMEM_SIZE`, around **60MB**.

However, my TA gets errors at the segmentation fault. Could you please tell me, do I understand the memory map size in the correct way? How to correctly increase the secure memory?

Thank you in advance.
Vincent

**Comment by pangupta** [comment]
Can you share the board name, you are using?
Can you share the DRAM size on the board?

Also share the crash logs.

**Comment by mofanv** [comment]
@pangupta Thanks for the reply.
The board is [Grapeboard](https://www.grapeboard.com/). The DRAM size is 1GB.
The OP-TEE Implementation is from [Openenclave](https://github.com/Microsoft/openenclave/blob/feature.new_platforms/new_platforms/docs/grapeboard.md).

I cannot see the crash logs, but only the `segmentation fault`. Even through having set the `CFG_TEE_CORE_LOG_LEVEL=3` and `CFG_TEE_TA_LOG_LEVEL=3`, I still couldn't see any `IMSG` and `DMSG` logs from the Serial terminal.

**Comment by pangupta** [comment]
I can see that you are trying to change parameters `PGT_CACHE_SIZE`, `TTBCR_N_VALUE`, ?
What is the reason for the same?

Yes, you can increase the OPTEE_OS DRAM size "CFG_TEE_OS_DRAM0_SIZE" by increasing its value.
and it should work, but we did not tested more than 64 MB.

**Comment by mofanv** [comment]
I changed `PGT_CACHE_SIZE` and `TTBCR_N_VALUE` according to steps in #2079. Before that, I have used this in QEMU.

I still think that the error in Grapeboard is because of the `TZDRAM` size. I tried several `TA_STACK_SIZE` and `TA_DATA_SIZE` in the Hello World example. After `TA_DATA_SIZE` larger than `5MB`, the app failed with `0xffff0008`. 

Do I miss some parameters to use larger secure memory? 5 MB is still far from 60 MB.
